File: UniverseUsingSTL.cpp
#include "UniverseUsingSTL.h"
#include <time.h>
#include <thread>
#include <chrono>
#include <stdlib.h>

using namespace std;

UniverseUsingSTL::UniverseUsingSTL(int amountOfRows,int amountOfColumns)
{
	this->setAmountOfRows(amountOfRows);
	this->setAmountOfColumns(amountOfColumns);
}

void UniverseUsingSTL::killCreature(Creature* C)
{
	bool found=false;
	int size=CreatureList.size();
	int i=0;
	while ((!found)&&(i<size))
	{
		if (CreatureList[i]==C)
		{
			found=true;
			swap(CreatureList[i],CreatureList[size-1]);
		}
		i++;
	}
	CreatureList.pop_back();
}

void UniverseUsingSTL::addCreature(Creature* C)
{
    CreatureList.push_back(C);
}

void UniverseUsingSTL::addRandomCreature(int amount)
{
	Creature* temp;
	srand(time(NULL));
	//randomly create creature
	for (int i = 0; i < amount; i++){
		int r;
		int row;
		int column;
		int directionX;
		int directionY;
		r = (rand() % 3);
		row = (rand() % getAmountOfRows());
		column = (rand() % getAmountOfColumns());
		directionX = (rand() % 2);
		if (directionX == 0){
			directionX = -1;
		}
		directionY = (rand() % 2);
		if (directionY == 0){
			directionY = -1;
		}
		int Found = 1;
        int counter = 0;

        //cari apakah sudah ada
        while((counter<(getAmountOfColumns()*getAmountOfRows()))&&(Found)){
            Found = 0;
            counter = 0;
            int sz = CreatureList.size();
            int i = 0;
            while((i<sz) && (!Found)&& (counter<(getAmountOfColumns()*getAmountOfRows())) ){
                counter++;
                if((CreatureList[i]->getRowPosition() == row) && (CreatureList[i]->getColumnPosition() == column))
                    Found = 1;
                i++;
            }

            if(Found)
            {
                column++;
                if (column == getAmountOfColumns()){column = 0; row++;}
                if (row == getAmountOfRows()){column = 0; row =0;}
            }

        }
        if(counter<(getAmountOfColumns()*getAmountOfRows())){
			if (r == 0){
				temp = new Plant(row, column);
				addCreature((temp));
			}
			else if (r == 1){
				temp = new Lamia(row, column, directionX, directionY);
				addCreature((temp));
			}
			else if (r == 2){
				temp = new Centaur(row, column, directionX, directionY);
				addCreature((temp));
			}
			else{
				temp = new Harpy(row, column, directionX, directionY);
				addCreature((temp));
			}
		}
	}
}

int UniverseUsingSTL::isWorldEmpty()
{
	return (CreatureList.size()==0);
}

void UniverseUsingSTL::checkForCollisions()
{
	int sz = CreatureList.size();
	int i = 0;
	//Kill Out Of Bounds
	while (i<sz)
	{
		if ((CreatureList[i]->getRowPosition()<0)||
			(CreatureList[i]->getColumnPosition()<0)||
			(CreatureList[i]->getRowPosition()>=getAmountOfRows())||
			(CreatureList[i]->getColumnPosition()>=getAmountOfColumns()))
		{
			killCreature(CreatureList[i]);
			i--;
			sz--;
		}
		i++;
	}

	//Kill Collisions
	sz = CreatureList.size();
	for (i=0;i<getAmountOfRows();i++)
	{
		for (int j=0;j<getAmountOfColumns();j++)
		{
			int minimumStrength = 999; //UNDEF
			int minimumIndex = -1;
			int counter = 0;
			for (int k=0;k<sz;k++)
			{
				if (((CreatureList[k])->getRowPosition()==i)&&
					((CreatureList[k])->getColumnPosition()==j))
				{
					if (minimumStrength>CreatureList[k]->getStrength())
					{
						minimumStrength=CreatureList[k]->getStrength();
						minimumIndex=k;
					}
					counter++;
				}
			}
			if (counter>1)
			{
				killCreature(CreatureList[minimumIndex]);
				sz--;
			}
		}
	}
}

void UniverseUsingSTL::moveAllCreaturesOnce()
{
    for (int i=0;i<CreatureList.size();i++)
    {
    	CreatureList[i]->doAction();
    }
    checkForCollisions();
}

void UniverseUsingSTL::print(ostream& output)
{
	char board[getAmountOfRows()][getAmountOfColumns()];
	for (int i=0;i<getAmountOfRows();i++)
	{
		for (int j=0;j<getAmountOfColumns();j++)
		{
			board[i][j] = '.';
		}
	}

	for (int i=0;i<CreatureList.size();i++)
	{
		Creature* currentCreature = (CreatureList[i]);
		board[currentCreature->getRowPosition()][currentCreature->getColumnPosition()]=currentCreature->draw();
	}

	for (int i=0;i<getAmountOfRows();i++)
	{
		for (int j=0;j<getAmountOfColumns();j++)
		{
			output << board[i][j];
		}
		output << endl;
	}
}

void UniverseUsingSTL::createThreadsForCreatures(){
 	// int sz = CreatureList.size();
 	// int i = 0;
 	// thread t[sz];
  //   for (i;i<sz;i++){
  //  		t[i] = thread(activateCreature);
  //       i++;
  //   }
  //   i = 0;
  //   for(i=0;i<sz;i++){
  //       t[i].join();
  //   }

}

void UniverseUsingSTL::activateCreature(){
		// Creature* C = CreatureList[i];
  //       C->doAction();
  //       this_thread::sleep_for(chrono::milliseconds(C->getActionInterval()));
  //       checkForCollisions();

}

File: java/UniverseUsingSTL.java
import java.io.*;
import java.util.Random;
import java.util.Vector;
import java.util.Collections;


public class UniverseUsingSTL extends Universe {
  private Vector<Creature> CreatureList;

  public UniverseUsingSTL(int amountOfRows, int amountOfColumns) {
    this.setAmountOfRows(amountOfRows);
    this.setAmountOfColumns(amountOfColumns);
    CreatureList = new Vector<Creature>();
  }

  public void killCreature(Creature creature) {
    boolean found = false;
    int size = CreatureList.size();
    int index = 0;
    while ((!found) && (index < size)) {
      if (CreatureList.get(index) == creature) {
        found = true;
        Collections.swap(CreatureList, index, CreatureList.size() - 1);
      }
      index++;
    }
    CreatureList.remove(CreatureList.size() - 1);
  }

  public void addCreature(Creature creature) {
    CreatureList.add(creature);
  }

  public void addRandomCreature(int amount) {
    Creature temp;
    Random generator = new Random();
    //randomly create creature
    for (int i = 0; i < amount; i++) {
      int rand = generator.nextInt(3);
      int row = (generator.nextInt(getAmountOfRows()));
      int column = (generator.nextInt(getAmountOfColumns()));
      int directionX = (generator.nextInt(2));
      if (directionX == 0) {
        directionX = -1;
      }
      int directionY = (generator.nextInt(2));
      if (directionY == 0) {
        directionY = -1;
      }
      boolean found = true;
      int counter = 0;

      //cari apakah sudah ada
      while ((counter < (getAmountOfColumns() * getAmountOfRows())) && (found)) {
        found = false;
        counter = 0;
        int sz = CreatureList.size();
        int index = 0;
        while ((index < sz) && (!found) && (counter < (getAmountOfColumns()
            * getAmountOfRows())) ) {
          counter++;
          if ((CreatureList.get(index).getRowPosition() == row)
              && (CreatureList.get(index).getColumnPosition() == column)) {
            found = true;
          }
          index++;
        }
        if (found) {
          column++;
          if (column == getAmountOfColumns()) {
            column = 0;
            row++;
          }
          if (row == getAmountOfRows()) {
            column = 0;
            row = 0;
          }
        }
      }
      if (counter < (getAmountOfColumns() * getAmountOfRows())) {
        if (rand == 0) {
          temp = new Plant(row, column);
          addCreature((temp));
        } else if (rand == 1) {
          temp = new Lamia(row, column, directionX, directionY);
          addCreature((temp));
        } else if (rand == 2) {
          temp = new Centaur(row, column, directionX, directionY);
          addCreature((temp));
        } else {
          temp = new Harpy(row, column, directionX, directionY);
          addCreature((temp));
        }
      }
    }
  }

  public boolean isWorldEmpty() {
    return (CreatureList.size() == 0);
  }

  public void checkForCollisions() {
    int sz = CreatureList.size();
    int index = 0;
    //Kill Out Of Bounds
    while (index < sz) {
      if ((CreatureList.get(index).getRowPosition() < 0)
          || (CreatureList.get(index).getColumnPosition() < 0)
          || (CreatureList.get(index).getRowPosition() >= getAmountOfRows())
          || (CreatureList.get(index).getColumnPosition() >= getAmountOfColumns())) {
        killCreature(CreatureList.get(index));
        index--;
        sz--;
      }
      index++;
    }

    //Kill Collisions
    sz = CreatureList.size();
    for (int i = 0; i < getAmountOfRows(); i++) {
      for (int j = 0; j < getAmountOfColumns(); j++) {
        int minimumStrength = 999; //UNDEF
        int minimumIndex = -1;
        int counter = 0;
        for (int k = 0; k < sz; k++) {
          if (((CreatureList.get(k)).getRowPosition() == i)
                && ((CreatureList.get(k)).getColumnPosition() == j)) {
            if (minimumStrength > CreatureList.get(k).getStrength()) {
              minimumStrength = CreatureList.get(k).getStrength();
              minimumIndex = k;
            }
            counter++;
          }
        }
        if (counter > 1) {
          killCreature(CreatureList.get(minimumIndex));
          sz--;
        }
      }
    }
  }

  public void moveAllCreaturesOnce() {
    for (int i = 0; i < CreatureList.size(); i++) {
      CreatureList.get(i).doAction();
    }
    checkForCollisions();
  }

  public void print() {
    char[][] board = new char[getAmountOfRows()][getAmountOfColumns()];
    for (int i = 0; i < getAmountOfRows(); i++) {
      for (int j = 0; j < getAmountOfColumns(); j++) {
        board[i][j] = '.';
      }
    }
    for (int i = 0; i < CreatureList.size(); i++) {
      Creature currentCreature = (CreatureList.get(i));
      board[currentCreature.getRowPosition()][currentCreature.getColumnPosition()] =
          currentCreature.draw();
    }

    for (int i = 0; i < getAmountOfRows(); i++) {
      for (int j = 0; j < getAmountOfColumns(); j++) {
        System.out.print(board[i][j]);
      }
      System.out.println();
    }
  }

  public void printFile(String str) throws IOException {
    try {
      BufferedWriter outputWriter = new BufferedWriter(new OutputStreamWriter(
          new FileOutputStream(str)));
      char[][] board = new char[getAmountOfRows()][getAmountOfColumns()];
      for (int i = 0; i < getAmountOfRows(); i++) {
        for (int j = 0; j < getAmountOfColumns(); j++) {
          board[i][j] = '.';
        }
      }

      for (int i = 0; i < CreatureList.size(); i++) {
        Creature currentCreature = (CreatureList.get(i));
        board[currentCreature.getRowPosition()][currentCreature.getColumnPosition()] =
        currentCreature.draw();
      }

      for (int i = 0; i < getAmountOfRows(); i++) {
        for (int j = 0; j < getAmountOfColumns(); j++) {
          outputWriter.write(board[i][j]);
        }
        outputWriter.newLine();
      }
		outputWriter.flush();
    } catch (FileNotFoundException exception) {
      System.out.println("File tidak ditemukan");
      System.exit(0);
    }
  }
  
    public static void main(String args[]){
	  UniverseUsingSTL world= new UniverseUsingSTL(10,10);
	  
	  for (int i = 0 ; i < 10 ; i ++){
		  world.addRandomCreature(3);
		  world.print();		 
		  System.out.println();
	  }
	  try{
			world.printFile("tes.txt");
			System.out.println("Printed file to tes.txt");
		} catch (IOException e){
			System.out.println("File tidak ditemukan");
			System.exit(0);		
		}
	 }
 
}
